File: test/test/material/beam/fragments/ChlapikBeamFragments.java
package material.beam.fragments;

import static com.xenoage.utils.collections.CollectionUtils.alist;
import static com.xenoage.utils.math.Fraction.fr;
import static com.xenoage.zong.core.music.Pitch.pi;
import static com.xenoage.zong.core.music.beam.Beam.beamFromChordsUnchecked;

import java.util.List;

import com.xenoage.utils.math.Fraction;
import com.xenoage.zong.core.music.beam.Beam;
import com.xenoage.zong.core.music.chord.Chord;
import com.xenoage.zong.core.music.chord.Note;

/**
 * These test examples are from Chlapik, page 45, rule 6.
 * 
 * @author Andreas Wenger
 */
public class ChlapikBeamFragments {

	public Beam exampleRow1Col1() {
		List<Chord> chords = alist();
		chords.add(chordC(Companion.fr(1, 16)));
		chords.add(chordC(Companion.fr(1, 8)));
		chords.add(chordC(Companion.fr(1, 16)));
		return beamFromChordsUnchecked(chords);
	}

	public Beam exampleRow1Col2() {
		List<Chord> chords = alist();
		chords.add(chordF(Companion.fr(1, 16)));
		chords.add(chordF(Companion.fr(1, 8)));
		chords.add(chordF(Companion.fr(1, 16)));
		return beamFromChordsUnchecked(chords);
	}

	public Beam exampleRow1Col3() {
		List<Chord> chords = alist();
		chords.add(chordC(Companion.fr(1, 16)));
		chords.add(chordC(Companion.fr(1, 8)));
		chords.add(chordC(Companion.fr(1, 8)));
		chords.add(chordC(Companion.fr(1, 16)));
		return beamFromChordsUnchecked(chords);
	}

	public Beam exampleRow1Col4() {
		List<Chord> chords = alist();
		chords.add(chordF(Companion.fr(1, 16)));
		chords.add(chordF(Companion.fr(1, 8)));
		chords.add(chordF(Companion.fr(1, 8)));
		chords.add(chordF(Companion.fr(1, 16)));
		return beamFromChordsUnchecked(chords);
	}

	public Beam exampleRow2Col1() {
		List<Chord> chords = alist();
		chords.add(chordC(Companion.fr(3, 16)));
		chords.add(chordC(Companion.fr(1, 16)));
		return beamFromChordsUnchecked(chords);
	}

	public Beam exampleRow2Col2() {
		List<Chord> chords = alist();
		chords.add(chordF(Companion.fr(3, 16)));
		chords.add(chordF(Companion.fr(1, 16)));
		return beamFromChordsUnchecked(chords);
	}

	public Beam exampleRow2Col3() {
		List<Chord> chords = alist();
		chords.add(chordC(Companion.fr(3, 32)));
		chords.add(chordC(Companion.fr(1, 32)));
		chords.add(chordC(Companion.fr(3, 32)));
		chords.add(chordC(Companion.fr(1, 32)));
		return beamFromChordsUnchecked(chords);
	}

	public Beam exampleRow2Col4() {
		List<Chord> chords = alist();
		chords.add(chordF(Companion.fr(3, 32)));
		chords.add(chordF(Companion.fr(1, 32)));
		chords.add(chordF(Companion.fr(3, 32)));
		chords.add(chordF(Companion.fr(1, 32)));
		return beamFromChordsUnchecked(chords);
	}

	public Beam exampleRow2Col5() {
		List<Chord> chords = alist();
		chords.add(chordC(Companion.fr(1, 8)));
		chords.add(chordC(Companion.fr(1, 32)));
		chords.add(chordC(Companion.fr(3, 32)));
		return beamFromChordsUnchecked(chords);
	}

	public Beam exampleRow2Col6() {
		List<Chord> chords = alist();
		chords.add(chordF(Companion.fr(1, 8)));
		chords.add(chordF(Companion.fr(1, 32)));
		chords.add(chordF(Companion.fr(3, 32)));
		return beamFromChordsUnchecked(chords);
	}

	public Beam exampleRow3Col2() {
		List<Chord> chords = alist();
		chords.add(chordC(Companion.fr(1, 16)));
		chords.add(chordC(Companion.fr(1, 8)));
		chords.add(chordC(Companion.fr(1, 16)));
		chords.add(chordC(Companion.fr(1, 8)));
		return beamFromChordsUnchecked(chords);
	}

	public Beam exampleRow3Col4() {
		List<Chord> chords = alist();
		chords.add(chordC(Companion.fr(1, 8)));
		chords.add(chordC(Companion.fr(1, 16)));
		chords.add(chordC(Companion.fr(3, 16)));
		return beamFromChordsUnchecked(chords);
	}

	public Beam exampleRow3Col6() {
		List<Chord> chords = alist();
		chords.add(chordC(Companion.fr(7, 32)));
		chords.add(chordC(Companion.fr(1, 32)));
		chords.add(chordC(Companion.fr(3, 32)));
		chords.add(chordC(Companion.fr(1, 32)));
		return beamFromChordsUnchecked(chords);
	}

	private Chord chordC(Fraction duration) {
		return new Chord(new Note(pi(0, 0, 5)), duration);
	}

	private Chord chordF(Fraction duration) {
		return new Chord(new Note(pi(3, 0, 4)), duration);
	}

}

File: core/core-jvm/test/material/beam/fragments/ChlapikBeamFragments.kt
package material.beam.fragments

import com.xenoage.utils.math.Fraction
import com.xenoage.utils.math.Fraction.Companion.fr
import com.xenoage.zong.core.music.Pitch.Companion.pi
import com.xenoage.zong.core.music.Step.C
import com.xenoage.zong.core.music.Step.F
import com.xenoage.zong.core.music.beam.Beam
import com.xenoage.zong.core.music.chord.Chord

/**
 * These test examples are from Chlapik, page 45, rule 6.
 */
class ChlapikBeamFragments {

	fun exampleRow1Col1() = Beam(listOf(
			chordC(fr(1, 16)),
			chordC(fr(1, 8)),
			chordC(fr(1, 16))))

	fun exampleRow1Col2() = Beam(listOf(
			chordF(fr(1, 16)),
			chordF(fr(1, 8)),
			chordF(fr(1, 16))))

	fun exampleRow1Col3() = Beam(listOf(
			chordC(fr(1, 16)),
			chordC(fr(1, 8)),
			chordC(fr(1, 8)),
			chordC(fr(1, 16))))


	fun exampleRow1Col4() = Beam(listOf(
			chordF(fr(1, 16)),
			chordF(fr(1, 8)),
			chordF(fr(1, 8)),
			chordF(fr(1, 16))))

	fun exampleRow2Col1() = Beam(listOf(
			chordC(fr(3, 16)),
			chordC(fr(1, 16))))

	fun exampleRow2Col2() = Beam(listOf(
			chordF(fr(3, 16)),
			chordF(fr(1, 16))))

	fun exampleRow2Col3() = Beam(listOf(
			chordC(fr(3, 32)),
			chordC(fr(1, 32)),
			chordC(fr(3, 32)),
			chordC(fr(1, 32))))

	fun exampleRow2Col4() = Beam(listOf(
			chordF(fr(3, 32)),
			chordF(fr(1, 32)),
			chordF(fr(3, 32)),
			chordF(fr(1, 32))))

	fun exampleRow2Col5() = Beam(listOf(
			chordC(fr(1, 8)),
			chordC(fr(1, 32)),
			chordC(fr(3, 32))))

	fun exampleRow2Col6() = Beam(listOf(
			chordF(fr(1, 8)),
			chordF(fr(1, 32)),
			chordF(fr(3, 32))))

	fun exampleRow3Col2() = Beam(listOf(
			chordC(fr(1, 16)),
			chordC(fr(1, 8)),
			chordC(fr(1, 16)),
			chordC(fr(1, 8))))

	fun exampleRow3Col4() = Beam(listOf(
			chordC(fr(1, 8)),
			chordC(fr(1, 16)),
			chordC(fr(3, 16))))

	fun exampleRow3Col6() = Beam(listOf(
			chordC(fr(7, 32)),
			chordC(fr(1, 32)),
			chordC(fr(3, 32)),
			chordC(fr(1, 32))))

	private fun chordC(duration: Fraction): Chord =
			Chord(duration, pi(C, 5))

	private fun chordF(duration: Fraction): Chord =
			Chord(duration, pi(F, 4))

}
